petethepig/opentelemetry-collector
# Adds profiles as another signal

This PR is my attempt at bringing profiling signal to the collector. It depends on [proto defenitions from open-telemetry/opentelemetry-proto repo](https://github.com/open-telemetry/opentelemetry-proto/pull/485). 

Most of this code is boilerplate, there's a handful of interesting files:
* `profiles.proto` in `opentelemetry-proto` repo ([link](https://github.com/open-telemetry/opentelemetry-proto/pull/485/files#diff-57ba5958fad9a734c920da5b8f3300ec33d66cada3ff3a807c024ba64bbf5948))
* `pdata/internal/cmd/pdatagen/internal/pprofile_package.go` connects everything to pdata and needs to be updated manually when proto changes ([link](https://github.com/petethepig/opentelemetry-collector/blob/da10c9b271d9696c9d37fa9c1383491d412100e7/pdata/internal/cmd/pdatagen/internal/pprofile_package.go#L6))
* `pdata/pprofile/json.go` implements json unmarshaling and needs to be updated manually when proto changes ([link](https://github.com/petethepig/opentelemetry-collector/blob/da10c9b271d9696c9d37fa9c1383491d412100e7/pdata/pprofile/json.go#L96))

---

Workflow I use to test this right now:
* Run `make genproto` to generate go bindings
  * (I replace proto version from git with a local one using `rsync -av ../opentelemetry-proto/ ${OPENTELEMETRY_PROTO_SRC_DIR}`)
* Run `make genpdata` to generate more pdata bindings
* Run `make otelcorecol` to verify that collector compiles
* Run `./bin/otelcorecol_darwin_arm64 --config=file:tmp/config.yaml` (config is [here](https://gist.github.com/petethepig/418a19b9a5b6f574598a23e93a2303be))
* Run `curl -i -X POST http://localhost:4318/v1/profiles -H 'Content-Type: application/json' -d @profile.json` (profile.json is [here](https://gist.github.com/petethepig/418a19b9a5b6f574598a23e93a2303be))
* ^ This prints the profile in text form

--- 

## To run benchmarks:

Clone this repo, make sure you're on the right branch (`profiling`). Run the following command:
```
make profile-benchmark
```

If you want to make changes to the proto:
* make sure you have [open-telemetry/opentelemetry-proto-profile](https://github.com/open-telemetry/opentelemetry-proto-profile) repo checkout out in an adjacent directory (`../opentelemetry-proto/`)
* The most recent version of that repo does not have alternative represenations, so you will need to checkout a specific commit: `git checkout 622c1658673283102a9429109185615bfcfaa78e`
* make changes to the profiling proto in `../opentelemetry-proto/`
* run `make genproto` to generate protobuf bindings
* run `make genpdata` to generate pdata bindings

You can now run benchmarks again:
```
make profile-benchmark
```

Benchmarks live in `/pdata/pprofile` directory. The files that you might want to look at are:
* `/pdata/pprofile/conversion.go`
* `/pdata/pprofile/benchmark_test.go`
